api/tests.py

```diff
@@ -71,3 +71,83 @@ def test_options_manifest_data(self):
+        from api.management.commands.importer import ImportSpec
+        from api.management.commands.importer import ImportOptions
+        from pathlib import Path
+        import json
+        
+        self.test_document_json = """
+            {
+            "title": "Test Reference Document",
+            "slug": "test-doc",
+            "desc": "This is a test document",
+            "license": "Open Gaming License",
+            "author": "John Doe",
+            "organization": "Open5e Test Org",
+            "version": "9.9",
+            "copyright": "",
+            "url": "http://example.com"
+            }
+        """
+        
+        self.test_spell_json = """
+            {
+            "name": "Magic Missile",
+            "desc": "You create three glowing darts of magical force. Each dart hits a creature of your choice that you can see within range. A dart deals 1d4 + 1 force damage to its target. The darts all strike simultaneously, and you can direct them to hit one creature or several.",
+            "higher_level": "When you cast this spell using a spell slot of 2nd level or higher, the spell creates one more dart for each slot level above 1st.",
+            "page": "phb 257",
+            "range": "120 feet",
+            "components": "V, S",
+            "ritual": "no",
+            "duration": "Instantaneous",
+            "concentration": "no",
+            "casting_time": "1 action",
+            "level": "1st-level",
+            "level_int": 1,
+            "school": "Evocation",
+            "class": "Sorcerer, Wizard"
+            }
+        """
+        i = Importer(ImportOptions(update=True, append=False, testrun=False))
+        i.import_document(json.loads(self.test_document_json),ImportSpec("test_filename","test_model_class","import_spell"))
+        i.import_spell(json.loads(self.test_spell_json),ImportSpec("test_filename","test_model_class","import_spell"))
+
+    def test_get_spells(self):
+        response = self.client.get(f'/spells/?format=json')
+        # Confirm the basic elements show up.
+        self.assertContains(response,'count',count=1)
+        self.assertContains(response,'next',count=1)
+        self.assertContains(response,'previous',count=1)
+        self.assertContains(response,'results',count=1)
+
+    def test_get_spell_data(self):
+        import json
+        response = self.client.get(f'/spells/?format=json')
+        # Confirm the basic elements show up.
+        self.assertContains(response,'count',count=1)
+        self.assertContains(response,'next',count=1)
+        self.assertContains(response,'previous',count=1)
+        self.assertContains(response,'results',count=1)
+
+        # Confirm the values appear as expected.
+        self.assertEqual(response.json()['count'],1)
+        in_mm = json.loads(self.test_spell_json)
+        out_mm = response.json()['results'][0]
+        
+        self.assertEqual(in_mm['name'], out_mm['name'])
```

Comment:
Could you put these assertions in a loop? Like:

```py
equal_fields = ('desc', 'higher_level', 'page', 'range', ...)
for field_name in equal_fields:
    self.assertEqual(in_mm[field_name] == out_mm[field_name])
self.assertEqual(in_mm['class'], out_mm['dnd_class'])
```

That will make the test more succinct, and thus easier to read and modify.

Reply:
That approach didn't seem to work for me, i was getting a range error, and I realized halfway through that this would result in terminal error messages that would show the VALUE, (i.e. AssertionError 1 !=0) but not have a clue on the NAME of the failed field, so I left it as-is after a bit of cleanup.

I like the idea of succinctness, but not at the cost of troubleshooting context.

Reply:
This is what the assert methods' `msg` kwarg is for. https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertEqual

> All the assert methods accept a msg argument that, if specified, is used as the error message on failure (see also [longMessage](https://docs.python.org/3/library/unittest.html#unittest.TestCase.longMessage)). Note that the msg keyword argument can be passed to [assertRaises()](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertRaises), [assertRaisesRegex()](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertRaisesRegex), [assertWarns()](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertWarns), [assertWarnsRegex()](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertWarnsRegex) only when they are used as a context manager.

So in this case, it would be like:

```py
self.assertEqual(in_mm[field_name], out_mm[field_name], f'Mismatched field {field_name}')
```

Reply:
Beautiful. More commits incoming.